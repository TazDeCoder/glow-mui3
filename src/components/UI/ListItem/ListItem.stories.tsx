import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { ListItem } from './ListItem';

export default {
  title: 'Components/UI/ListItem',
  component: ListItem,
} as ComponentMeta<typeof ListItem>;

export const Default: ComponentStory<typeof ListItem> = (args) => (
  <ListItem width={320} {...args}>
    Headline
  </ListItem>
);
