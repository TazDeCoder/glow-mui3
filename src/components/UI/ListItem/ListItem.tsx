/** @jsxRuntime classic */
/** @jsx jsx */

import { forwardRef, useState } from 'react';
import { jsx, ThemeUICSSObject, ThemeUIStyleObject } from '@theme-ui/core';

import { Box, BoxProps, BoxOwnProps } from '../../Utils';
import { Typography } from '../Typography';
import { __internalProps } from '../../shared/utils';
import type { Assign, ForwardRef } from '../../types';

export interface ListItemProps
  extends Assign<React.ComponentPropsWithRef<'li'>, BoxOwnProps> {
  /** Disabled state of the component
   * @default false
   */
  disabled?: boolean;
}

const __itemCss: ThemeUICSSObject = {
  position: 'relative',
  display: 'flex',
  alignItems: 'center',
  height: 56,
  pr: 24,
  py: 2,
  bg: 'surface.main',
  borderRadius: 'none',
  '&:nth-last-of-type(n + 1)': {
    pl: 3,
  },
  '&:focus::after': {
    content: '""',
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    bg: 'action.active',
    borderRadius: 'inherit',
  },
};

export const ListItem: ForwardRef<HTMLLIElement, ListItemProps> = forwardRef(
  ({ children, disabled = false, sx, ...props }, ref) => {
    const [isDrag, setIsDrag] = useState(false);

    const style: ThemeUIStyleObject = {
      ...(!disabled
        ? {
          cursor: 'pointer',
          '&:hover::after': {
            content: '""',
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            bg: 'action.hover',
            borderRadius: 'inherit',
          },
          '&:active::after': {
            content: '""',
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            bg: isDrag ? 'action.drag' : 'action.active',
            borderRadius: 'inherit',
            color: 'surface.variant',
          },
        }
        : {
          bg: 'surface.main',
          opacity: 0.38,
          userSelect: 'none',
        }),
    };

    const dragHandler = () => {
      setIsDrag((prevState) => !prevState);
    };

    return (
      <Box
        ref={ref}
        as="li"
        sx={{ listStyle: 'none', ...sx }}
        {...(props as BoxProps)}
      >
        <Box
          sx={style}
          draggable={!disabled}
          onDragEnter={dragHandler}
          onDragLeave={dragHandler}
          {...__internalProps({ __css: __itemCss })}
        >
          <Typography color="surface.onMain">{children}</Typography>
        </Box>
      </Box>
    );
  },
);
