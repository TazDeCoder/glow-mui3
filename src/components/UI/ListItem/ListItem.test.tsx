import { render } from '@testing-library/react';

import { ListItem } from './ListItem';

describe('<ListItem />', () => {
  describe('Snapshots', () => {
    it('renders correctly', () => {
      const { container } = render(<ListItem>Foo</ListItem>);
      expect(container).toMatchSnapshot();
    });
  });
});
