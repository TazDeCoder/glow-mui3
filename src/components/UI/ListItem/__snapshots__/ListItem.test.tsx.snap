// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem /> Snapshots renders correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  list-style: none;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 56px;
  padding-right: 24px;
  padding-top: 8px;
  padding-bottom: 8px;
  background-color: surface.main;
  border-radius: none;
  cursor: pointer;
}

.emotion-1:nth-last-child(n + 1) {
  padding-left: 16px;
}

.emotion-1:focus::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: action.active;
  border-radius: inherit;
}

.emotion-1:hover::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: action.hover;
  border-radius: inherit;
}

.emotion-1:active::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: action.active;
  border-radius: inherit;
  color: surface.variant;
  z-index: 1;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: surface.onMain;
}

<div>
  <li
    class="emotion-0"
  >
    <div
      class="emotion-1"
      draggable="true"
    >
      <p
        class="emotion-2"
      >
        Foo
      </p>
    </div>
  </li>
</div>
`;
